aleph/views/mappings_api.py

```diff
@@ -0,0 +1,100 @@
+import logging
+from banal import first
+from followthemoney import model
+from flask import Blueprint, request
+from werkzeug.exceptions import BadRequest
+
+from aleph.model import Mapping
+from aleph.search import QueryParser, DatabaseQueryResult
+from aleph.views.serializers import MappingSerializer
+from aleph.views.util import get_db_collection, get_index_entity, parse_request  # noqa
+from aleph.views.forms import MappingSchema
+from aleph.views.util import require, obj_or_404
+from aleph.logic.mapping import load_mapping, flush_mapping
+
+blueprint = Blueprint('mappings_api', __name__)
+log = logging.getLogger(__name__)
+
+
+def load_query():
+    try:
+        query = request.json.get('mapping_query', '{}')
+        # just for validation
+        model.make_mapping({'entities': query})
+    except Exception as ex:
+        raise BadRequest(ex)
+    return query
+
+
+@blueprint.route('/api/2/collections/<int:collection_id>/mappings', methods=['GET'])  # noqa
+def index(collection_id):
+    collection = get_db_collection(collection_id)
+    parser = QueryParser(request.args, request.authz)
+    table_id = first(parser.filters.get('table'))
+    q = Mapping.by_collection(collection.id, table_id=table_id)
+    result = DatabaseQueryResult(request, q, parser=parser)
+    return MappingSerializer.jsonify_result(result)
+
+
+@blueprint.route('/api/2/collections/<int:collection_id>/mappings', methods=['POST', 'PUT'])  # noqa
+def create(collection_id):
+    collection = get_db_collection(collection_id, request.authz.WRITE)
+    data = parse_request(MappingSchema)
+    entity_id = data.get('table_id')
+    query = load_query()
+    entity = get_index_entity(entity_id, request.authz.READ)
+    mapping = Mapping.create(query, entity.get('id'), collection, request.authz.id)  # noqa
+    return MappingSerializer.jsonify(mapping)
+
+
+@blueprint.route('/api/2/collections/<int:collection_id>/mappings/<int:mapping_id>', methods=['GET'])  # noqa
+def view(collection_id, mapping_id):
+    get_db_collection(collection_id, request.authz.WRITE)
+    mapping = obj_or_404(Mapping.by_id(mapping_id))
+    return MappingSerializer.jsonify(mapping)
+
+
+@blueprint.route('/api/2/collections/<int:collection_id>/mappings/<int:mapping_id>', methods=['POST', 'PUT'])  # noqa
+def update(collection_id, mapping_id):
+    collection = get_db_collection(collection_id, request.authz.WRITE)
+    mapping = obj_or_404(Mapping.by_id(mapping_id))
+    data = parse_request(MappingSchema)
+    entity_id = data.get('table_id')
+    query = load_query()
+    entity = get_index_entity(entity_id, request.authz.READ)
+    mapping.update(query=query, table_id=entity.get('id'))
+    if request.args.get('flush') == 'true':
```

Comment:
There are two ways to trigger flush and load, both as a query argument and a separate endpoint. I prefer the separate endpoint and would get rid of the flags. Then we can also move the logic of `flush_mapping` and `load_mapping` into the view functions more easily.